File: chapter1/1.4/python/solution_test.py
import pytest

from itertools import permutations

from solution import palindrome_permutations, custom_permutations


@pytest.mark.parametrize(
    "input_str,expected",
    [
        ("", set([""])),
        ("aa", set(["aa"])),
        ("aba", set(["aab", "aba", "baa"])),
        ("abc", set()),
    ],
)
def test_palindrome_permutations(input_str: str, expected: tuple):
    assert palindrome_permutations(input_str) == expected


@pytest.mark.parametrize("input_str", [("abc",), ("abad",), ("llama",)])
def test_custom_permutations(input_str: str):
    custom_permutations_set = set("".join(x) for x in custom_permutations(input_str))
    standard_permutations_set = set("".join(x) for x in permutations(input_str))
    assert standard_permutations_set == custom_permutations_set

File: chapter1/1.4/go/solution_test.go
package main

import (
	"reflect"
	"testing"
)

func TestIsPermutation(t *testing.T) {
	var testCases = []struct {
		s        string
		expected []string
	}{
		{"", []string{""}},
		{"abc", nil},
		{"aab", []string{"aab", "aba", "baa"}},
	}
	for _, tc := range testCases {
		t.Run(tc.s, func(t *testing.T) {
			palindromes := PalindromePermutations(tc.s)
			if reflect.DeepEqual(palindromes, tc.expected) {
				t.Errorf("PalindromePermutations(%v) should return %[2]v(%[2]T) but got %[3]v(%[3]T)", tc.s, tc.expected, palindromes)
			}
		})
	}
}
